### PR TITLE
Fix NOT-IN index optimization when not all key columns are used.

### LINKED ISSUE #28120: Invalid NOT IT statement behaviour for empty(1970-01-01) date
**Describe what's wrong**

Since version 21.7 was introduced [NOT IN](https://github.com/ClickHouse/ClickHouse/pull/24894) optimization 
that does not work correctly for expression with an empty(default)  dates

**Does it reproduce on recent release?**

v21.7.2.7-stable

**How to reproduce**

```
create table t1 (date Date, a Float64, b String) Engine=MergeTree ORDER BY date
create table t2 (date Date, a Float64, b String) Engine=MergeTree ORDER BY date

insert into t1(a, b) values (1, 'one'), (2, 'two')
insert into t2(a, b) values (2, 'two'), (3, 'three')

```

Request like 
```
select date, a, b
from t1
where (date, a, b) NOT IN (select date,a,b from t2)
```

or 

```
select date, a, b
from t2
where (date, a, b) NOT IN (select date,a,b from t1)
```

does not return anything

**Expected behavior**

The first request should returns `│ 1970-01-01 │ 1 │ one │` .
The second request should returns `│ 1970-01-01 │ 3 │ three │`.

**Additional context**

If write statement without date field it works as expected:

```
select date, a, b
from t1
where (a, b) NOT IN (select a,b from t2)
```
returns  `│ 1970-01-01 │ 1 │ one │`

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix NOT-IN index optimization when not all key columns are used. This fixes https://github.com/ClickHouse/ClickHouse/issues/28120

Detailed description / Documentation draft:
.

### CODE REVIEW

File: src/Interpreters/Set.cpp [hunk 1]
```diff
@@ -402,8 +402,8 @@ void Set::checkTypesEqual(size_t set_type_idx, const DataTypePtr & other_type) c
                         + data_types[set_type_idx]->getName() + " on the right", ErrorCodes::TYPE_MISMATCH);
 }
 
-MergeTreeSetIndex::MergeTreeSetIndex(const Columns & set_elements, std::vector<KeyTuplePositionMapping> && index_mapping_)
-    : indexes_mapping(std::move(index_mapping_))
+MergeTreeSetIndex::MergeTreeSetIndex(const Columns & set_elements, std::vector<KeyTuplePositionMapping> && indexes_mapping_)
+    : all_key(set_elements.size() == indexes_mapping_.size()), indexes_mapping(std::move(indexes_mapping_))
```
Comment (novikd) [inline src/Interpreters/Set.cpp:406]: Let's initialize each member on a separate line

File: src/Interpreters/Set.h [hunk 2]
```diff
@@ -217,6 +217,8 @@ class MergeTreeSetIndex
     BoolMask checkInRange(const std::vector<Range> & key_ranges, const DataTypes & data_types) const;
 
 private:
+    // If all arguments in tuple are key columns, we can optimize NOT IN when there is only one element.
+    bool all_key;
```
Comment (novikd) [inline src/Interpreters/Set.h:221]: I think it's better to name this variable something like `maps_all_keys`

**Review by novikd — APPROVED** [review]

File: src/Interpreters/Set.cpp (hunk 1 above)
Comment (amosbird) [reply src/Interpreters/Set.cpp:406]: Interesting. I though we need to obey what clang-format says for new code.

File: src/Interpreters/Set.h (hunk 2 above)
Comment (amosbird) [reply src/Interpreters/Set.h:221]: hmm, but why maps? How about `has_all_keys`?

Comment (novikd) [reply src/Interpreters/Set.h:221]: That sounds good too.

File: src/Interpreters/Set.cpp (hunk 1 above)
Comment (novikd) [reply src/Interpreters/Set.cpp:406]: I think in that case this affects readability of the code. Also you can find similar code in the code base.

Comment (amosbird) [reply src/Interpreters/Set.cpp:406]: But we have a lot of these styles https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Set.h#L35

**Pull request merged**